### PR TITLE
Use vectors in managed ClientWebSocket for mask application

### PR DESCRIPTION
**Pull request by stephentoub**
The web socket protocol uses a 4-byte mask sent from client to server and xor'd with the payload data.  In the managed web socket implementation, that mask is currently applied byte-by-byte.  This commit changes it to be done using SIMD.

Micro-benchmarks on the ApplyMask function itself show a regression for payload lengths under 4 bytes, but above that it's a clear win:
```
1: 0.44x
3: 0.64x
4: 1.15x
7: 1.16x
31: 1.73x
32: 6.06x
63: 3.37x
64: 10.40x
1024: 20.36x
```
This doesn't have a measurable affect on small payloads in the overall context of sending/receiving data, but the larger the payload, the more it applies.  Locally I measured upwards of a 10% improvement for payloads several-K in size.

cc: @benaadams, @geoffkizer, @davidsh, @anurse ,@mellinoe

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please
@dotnet-bot test Outerloop Ubuntu14.04 Debug please
@dotnet-bot test Outerloop OSX Debug please

File: src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 1]
```diff
@@ -1202,21 +1203,69 @@ private static int ApplyMask(byte[] toMask, int toMaskOffset, byte[] mask, int m
...
         {
-            Debug.Assert(toMaskOffset <= toMask.Length - count, $"Unexpected inputs: {toMaskOffset}, {toMask.Length}, {count}");
-            Debug.Assert(maskIndex < sizeof(int), $"Unexpected {nameof(maskIndex)}: {maskIndex}");
+            int maskShift = maskIndex * 8;
+            int shiftedMask = (int)(((uint)mask >> maskShift) | ((uint)mask << (32 - maskShift)));
+
+            // Try to use SIMD.  We can if the number of bytes we're trying to mask is at least as much
+            // as the width of a vector and if the width is an even multiple of the mask.
+            if (Vector.IsHardwareAccelerated &&
+                Vector<byte>.Count % sizeof(int) == 0 &&
```
Comment (benaadams) [inline src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs]: O_o will this be const elided by jit?

File: src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 2]
```diff
@@ -1202,21 +1203,69 @@ private static int ApplyMask(byte[] toMask, int toMaskOffset, byte[] mask, int m
...
+                count >= Vector<byte>.Count)
+            {
+                // Mask bytes a vector at a time.
+                Vector<byte> maskVector = Vector.AsVectorByte(new Vector<int>(shiftedMask));
+                while (count >= Vector<byte>.Count)
+                {
+                    (maskVector ^ new Vector<byte>(toMask, toMaskOffset)).CopyTo(toMask, toMaskOffset);
+                    toMaskOffset += Vector<byte>.Count;
+                    count -= Vector<byte>.Count;
```
Comment (benaadams) [inline src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs]: Can move this up two lines, immediately after `while` so its ready earlier in pipeline for next loop test

File: src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 3]
```diff
@@ -1202,21 +1203,69 @@ private static int ApplyMask(byte[] toMask, int toMaskOffset, byte[] mask, int m
...
+                    byte* p = toMaskPtr + toMaskOffset;
+
+                    // Try to go an int at a time if the remaining data is 4-byte aligned and there's enough remaining.
+                    if (((long)p % sizeof(int)) == 0)
+                    {
+                        while (count >= sizeof(int))
+                        {
+                            *((int*)p) ^= shiftedMask;
+                            p += sizeof(int);
+                            count -= sizeof(int);
```
Comment (benaadams) [inline src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs]: Can move this up two lines, immediately after while so its ready earlier in pipeline for next loop test

**Comment by benaadams** [comment]
LGTM

File: src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs (hunk 1 above)
Comment (stephentoub) [inline src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs]: Yup:
image

File: src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs (hunk 2 above)
Comment (stephentoub) [inline src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs]: Moved

File: src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs (hunk 3 above)
Comment (stephentoub) [inline src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs]: Moved

**Pull request merged**